### PR TITLE
Remove `seedTemplate` field from ManagedSeed API

### LINKED ISSUE #6898: Consider removing `SeedTemplate` field from ManagedSeed
**How to categorize this issue?**

/area scalability
/kind cleanup

**What would you like to be added**:
The field `SeedTemplate` in the ManagedSeed-API is used to register the `seed` object via the ManagedSeed-Controller rather than the Gardenlet.

```
	// SeedTemplate is a template for a Seed object, that should be used to register a given cluster as a Seed.
	// Either SeedTemplate or Gardenlet must be specified. When Seed is specified, the ManagedSeed controller will not deploy a gardenlet into the cluster
	// and an existing gardenlet reconciling the new Seed is required.
```

At the same time, the Gardenlet-Config **requires** the `config.seedConfig` to be set, i.e. it is anyway registering or updating the `seed` object when it starts.

The two concepts are conflicting and we should only follow one approach. Since Gardenlet mimics the behavior of Kubelet WRT node registration, it's preferred to get rid of the mentioned ManagedSeed use-case and to remove the `ManagedSeed.SeedTemplate` field completely.

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/kind api-change
/kind cleanup

**What this PR does / why we need it**:
This PR removes `managedSeed.spec.seedTemplate` after it has been deprecated in #7006.

**Which issue(s) this PR fixes**:
Fixes #6898

**Release note**:

```breaking operator
The deprecated field `managedSeed.spec.seedTemplate` has been removed from the ManagedSeed API. Please check your `ManagedSeed`s and `ManagedSeedSet`s and remove any usage (switch to `spec.gardenlet.config`) before upgrading to this Gardener version.
```

### CODE REVIEW

**Comment by timuthy** [comment]
As discussed, this PR will be transitioned to draft state since for the very next Gardener release we want to continue with a "softer" deprecation. See #7006.
In addition, this PR will be reworked so that it also contains the removal of the field.

**Comment by timuthy** [comment]
Update: PR has been reworked and can be released with `v1.61` at the earliest.
/cc @acumino to not add release milestone

File: pkg/apis/seedmanagement/v1alpha1/types_managedseed.go [hunk 1]
```diff
@@ -67,12 +67,10 @@ type ManagedSeedSpec struct {
...
 	Shoot *Shoot `json:"shoot,omitempty" protobuf:"bytes,1,opt,name=shoot"`
-	// SeedTemplate is a template for a Seed object, that should be used to register a given cluster as a Seed.
-	// Either SeedTemplate or Gardenlet must be specified. When Seed is specified, the ManagedSeed controller will not deploy a gardenlet into the cluster
-	// and an existing gardenlet reconciling the new Seed is required.
-	// Deprecated: This field is deprecated and will be removed in a future version of Gardener. Define seed via `.spec.gardenlet.config` instead.
-	// +optional
-	SeedTemplate *gardencorev1beta1.SeedTemplate `json:"seedTemplate,omitempty" protobuf:"bytes,2,opt,name=seedTemplate"`
+
+	// SeedTemplate is tombstoned to show why 2 is reserved protobuf tag.
+	// SeedTemplate *gardencorev1beta1.SeedTemplate `json:"seedTemplate,omitempty" protobuf:"bytes,2,opt,name=seedTemplate"`
```
Comment (timuthy) [inline pkg/apis/seedmanagement/v1alpha1/types_managedseed.go:72]: FYI, best practice following this [example](https://github.com/kubernetes/kubernetes/pull/102412/files#diff-5ffbebb40a29f8e491f4ff1e07c618eb3edcc0d5f00a1a71445272d6767d7883L4258). Thanks @ialidzhikov for the reference.

Comment (ialidzhikov) [reply pkg/apis/seedmanagement/v1alpha1/types_managedseed.go:72]: Thanks for incorporating my feedback. We could also add an unit test similar to the one added in https://github.com/kubernetes/kubernetes/pull/102412 to make sure that any new field won't reuse the protobuf number of the already removed field.

```go
package v1alpha1_test

import (
	"reflect"
	"strings"

	. "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"

	. "github.com/onsi/ginkgo/v2"
)

var _ = Describe("ManagedSeed", func() {

	It("should not allow reuse of protobuf number of already removed ManagedSeedSpec fields", func() {
		obj := reflect.ValueOf(ManagedSeedSpec{}).Type()
		for i := 0; i < obj.NumField(); i++ {
			f := obj.Field(i)

			protobufNum := strings.Split(f.Tag.Get("protobuf"), ",")[1]
			if protobufNum == "2" {
				Fail("protobuf 2 in ManagedSeedSpec is reserved for removed seedTemplate field")
			}
		}
	})
})
```

I also have in my list to update the api conventions doc in g/g wrt to this.

File: pkg/apis/seedmanagement/helper/helper_test.go [hunk 2]
```diff
@@ -97,7 +80,28 @@ var _ = Describe("Helper", func() {
...
+			})
+
+			It("gardenlet config is not defined", func() {
+				managedSeed.Spec.Gardenlet = &seedmanagement.Gardenlet{}
+
+				_, err := ExtractSeedSpec(managedSeed)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("seedConfig is not defined in gardenlet config", func() {
```
Comment (ialidzhikov) [inline pkg/apis/seedmanagement/helper/helper_test.go]: Nit: Just for completeness, is it possible to add a test case for the case when the gardenlet config fails to be decoded?

File: pkg/apis/seedmanagement/helper/helper.go [hunk 3]
```diff
@@ -32,19 +32,18 @@ func GetBootstrap(bootstrap *seedmanagement.Bootstrap) seedmanagement.Bootstrap
...
+		return nil, fmt.Errorf("no gardenlet config specified in managedseed %s", managedSeed.Name)
+	}
+
+	gardenletConfig, err := confighelper.ConvertGardenletConfiguration(gardenlet.Config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert gardenlet config for managedseed %s: %w", managedSeed.Name, err)
 	}
-	return nil, fmt.Errorf("no seed spec found for managedseed %s", managedSeed.Name)
+	if gardenletConfig.SeedConfig == nil {
+		return nil, fmt.Errorf("no seed spec found for managedseed %s", managedSeed.Name)
```
Comment (ialidzhikov) [inline pkg/apis/seedmanagement/helper/helper.go]: Nit:

```suggestion
		return nil, fmt.Errorf("no seed config found for managedseed %s", managedSeed.Name)
```

File: pkg/apis/seedmanagement/v1alpha1/types_managedseed.go (hunk 1 above)
Comment (timuthy) [reply pkg/apis/seedmanagement/v1alpha1/types_managedseed.go:72]: Thanks @ialidzhikov, added in [dd6510a](https://github.com/gardener/gardener/pull/6972/commits/dd6510aaf6633b76ee12086755d69cdd5373aa30).

**Review by ialidzhikov — COMMENTED** [review]
Minor suggestions otherwise lgtm

**Comment by timuthy** [comment]
I addressed your review comments @ialidzhikov, PTAL.

**Review by ialidzhikov — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**